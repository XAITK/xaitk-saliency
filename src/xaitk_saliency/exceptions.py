class ShapeMismatchError(Exception):
    """Exception for when matrix shape expectations are violated."""


class MismatchedLabelsError(Exception):
    """Raised when two sets of detections do not have matching class labels."""

    def __init__(self, message: str = "Detections have mismatched class labels.") -> None:
        self.message = message
        super().__init__(self.message)
