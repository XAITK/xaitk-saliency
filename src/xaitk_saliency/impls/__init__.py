"""Module for all implementations of xaitk-saliency interfaces"""
