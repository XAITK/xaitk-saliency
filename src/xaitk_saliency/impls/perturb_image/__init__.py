"""Module for all implementations of the PerturbImage interface"""
