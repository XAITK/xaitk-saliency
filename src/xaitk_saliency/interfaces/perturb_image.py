import abc

import numpy as np
from smqtk_core import Plugfigurable


class PerturbImage(Plugfigurable):
    """
    Interface abstracting the behavior of taking a reference image and
    generating some number perturbations in the form of mask matrices
    indicating where perturbations should occur and to what amount.

    Implementations should impart no side effects upon the input image.
    """

    @abc.abstractmethod
    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        """
        Transform an input reference image into a number of mask matrices
        indicating the perturbed regions.

        Output mask matrix should be three-dimensional with the format
        [nMasks x Height x Width], sharing the same height and width to the
        input reference image.
        The implementing algorithm may determine the quantity of output masks
        per input image.
        These masks should indicate the regions in the corresponding perturbed
        image that have been modified.
        Values should be in the [0, 1] range, where a value closer to 1.0
        indicates areas of the image that are unperturbed.
        Note that output mask matrices may be of a floating-point type
        to allow for fractional perturbation.

        :param ref_image:
            Reference image to generate perturbations from.
        :return: Mask matrix with shape [nMasks x Height x Width].
        """

    def __call__(self, ref_image: np.ndarray) -> np.ndarray:
        """Alias for :meth:`.PerturbImage.perturb`."""
        return self.perturb(ref_image)
