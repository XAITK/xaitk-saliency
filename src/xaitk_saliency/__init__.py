import pkg_resources

from .interfaces.gen_classifier_conf_sal import GenerateClassifierConfidenceSaliency  # noqa: F401
from .interfaces.gen_descriptor_sim_sal import GenerateDescriptorSimilaritySaliency  # noqa: F401
from .interfaces.gen_detector_prop_sal import GenerateDetectorProposalSaliency  # noqa: F401
from .interfaces.gen_image_classifier_blackbox_sal import GenerateImageClassifierBlackboxSaliency  # noqa: F401
from .interfaces.gen_image_similarity_blackbox_sal import GenerateImageSimilarityBlackboxSaliency  # noqa: F401
from .interfaces.gen_object_detector_blackbox_sal import GenerateObjectDetectorBlackboxSaliency  # noqa: F401
from .interfaces.perturb_image import PerturbImage  # noqa: F401

# It is known that this will fail if SMQTK-Core is not "installed" in the
# current environment. Additional support is pending defined use-case-driven
# requirements.
__version__ = pkg_resources.get_distribution(__name__).version
