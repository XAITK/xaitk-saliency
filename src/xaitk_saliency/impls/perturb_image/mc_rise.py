"""Implementation of MC-RISE perturbation mask generation"""

from typing import Any, Optional
from typing_extensions import override

import numpy as np
from skimage.transform import resize
from smqtk_descriptors.utils import parallel_map

from xaitk_saliency.interfaces.perturb_image import PerturbImage


class MCRISEGrid(PerturbImage):
    """
    Based on Hatakeyama et. al:
    https://openaccess.thecvf.com/content/ACCV2020/papers/Hatakeyama_Visualizing_Color-wise_Saliency_of_Black-Box_Image_Classification_Models_ACCV_2020_paper.pdf
    """

    def __init__(
        self,
        n: int,
        s: int,
        p1: float,
        k: int,
        seed: int = None,
        threads: Optional[int] = 4,
    ) -> None:
        """
        Generate a set of random binary masks

        :param n: int
            Number of random masks used in the algorithm. E.g. 1000.
        :param s: int
            Spatial resolution of the small masking grid. E.g. 8.
            Assumes square grid.
        :param p1: float
            Probability of the grid cell being set to 1 (otherwise 0).
            This should be a float value in the [0, 1] range. E.g. 0.5.
        :param k: int
            Number of colors to use.
        :param seed: Optional[int]
            A seed to pass into the constructed random number generator to allow
            for reproducibility
        :param threads: int
            The number of threads to utilize when generating masks.
            If this is <=0 or None, no threading is used and processing
            is performed in-line serially.

        :raises: ValueError
            If p1 not in [0, 1].
        :raises: ValueError
            If k < 1.
        """
        if p1 < 0 or p1 > 1:
            raise ValueError(f"Input p1 value of {p1} is not within the expected [0,1] range.")

        if k <= 0:
            raise ValueError(f"Input k value of {k} is not within the expected >0 range.")

        self.n = n
        self.s = s
        self.p1 = p1
        self.k = k
        self.seed = seed
        self.threads = threads

        # Generate a set of random grids of small resolution
        rng = np.random.default_rng(seed)
        grid: np.ndarray = rng.random((n, s, s)) < p1

        indiv_color_masks = np.empty((self.k, self.n, self.s, self.s))
        for g_idx, g in enumerate(grid):
            g_shape = g.shape
            # Randomly choose fill colors for each pixel
            fill_mask = rng.integers(0, k, size=g_shape)
            for fill_idx in range(self.k):
                # Keep the masked pixels where the current idx (color) was selected
                indiv_color_mask = np.where(g == False, np.where(fill_mask == fill_idx, False, True), True)
                indiv_color_masks[fill_idx][g_idx] = indiv_color_mask

        self.grid = indiv_color_masks
        self.grid.astype("float32")

    @override
    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        """
        Warning: this implementation returns a different shape than typically expected by this interface.
        Instead of `[nMasks x Height x Width]`, masks of shape `[kColors x nMasks x Height x Width]`
        are returned.

        :param ref_image: np.ndarray
            Reference image to generate perturbations from.

        :return: np.ndarray
            Mask matrix with shape `[kColors x nMasks x Height x Width]`.
        """
        input_size = np.shape(ref_image)[:2]
        num_colors = self.k
        num_masks = self.n
        grid = self.grid
        s = self.s
        shift_rng = np.random.default_rng(self.seed)
        # Shape format: [H x W], Inherits from `input_size`
        cell_size = np.ceil(np.array(input_size) / s)
        up_size = (s + 1) * cell_size

        masks = np.empty((num_colors, num_masks, *input_size), dtype=grid.dtype)

        # Expanding index accesses for repetition efficiency.
        cell_h, cell_w = cell_size[:2]
        input_h, input_w = input_size[:2]

        # flake8: noqa
        def work_func(i_: int) -> np.ndarray:
            # Random shifts
            y = shift_rng.integers(0, cell_h)
            x = shift_rng.integers(0, cell_w)
            k_masks = np.empty((num_colors, input_h, input_w))
            for k_ in range(num_colors):
                k_masks[k_] = resize(
                    grid[k_, i_, ...],
                    up_size,
                    order=1,
                    mode="reflect",
                    anti_aliasing=False,
                )[
                    y : y + input_h,
                    x : x + input_w,
                ]
            return k_masks

        threads = self.threads
        if threads is None or threads < 1:
            for i in range(num_masks):
                masks[:, i, ...] = work_func(i)
        else:
            for i, m in enumerate(
                parallel_map(
                    work_func,
                    range(num_masks),
                    cores=threads,
                    use_multiprocessing=False,
                ),
            ):
                masks[:, i, ...] = m

        return masks

    @override
    def get_config(self) -> dict[str, Any]:
        return {
            "n": self.n,
            "s": self.s,
            "p1": self.p1,
            "k": self.k,
            "seed": self.seed,
            "threads": self.threads,
        }
