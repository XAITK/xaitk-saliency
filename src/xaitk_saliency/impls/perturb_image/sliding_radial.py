from typing import Any, Optional

import numpy as np
from scipy.ndimage.filters import gaussian_filter
from skimage.draw import ellipse

from xaitk_saliency.interfaces.perturb_image import PerturbImage


class SlidingRadial(PerturbImage):
    """
    Produce perturbation matrices generated by sliding a radial occlusion area
    with configured radius over the area of an image. When the two radius values
    are the same, circular masks are generated; otherwise, elliptical masks are
    generated. Passing sigma values will apply a Gaussian filter to the mask,
    blurring it. This results in a smooth transition from full occlusion in the
    center of the radial to no occlusion at the edge.

    Due to the geometry of sliding radials, if the stride given does not evenly
    divide the radial size along the applicable axis, then the result plane of
    values when summing the generated masks will not be even.

    Related, if the stride is set to be larger than the radial diameter, the
    resulting plane of summed values will also not be even, as there be
    increasingly long valleys of unperturbed space between masked regions.

    The generated masks are boolean if no blurring is used, otherwise the masks
    will be of floating-point type in the [0, 1] range.

    :param radius: The radius of the occlusion area in pixels as a tuple with
        format `(radius_y, radius_x)`.
    :param stride: The striding step in pixels for the center of the radial as
        a tuple with format `(height_step, width_step)`.
    :param sigma: The sigma values for the Gaussian filter applied to masks in
        pixels as a tuple with format `(sigma_y, sigma_x)`.
    """

    def __init__(
        self,
        radius: tuple[float, float] = (50, 50),
        stride: tuple[int, int] = (20, 20),
        sigma: Optional[tuple[float, float]] = None,
    ) -> None:
        self.radius = (radius[0], radius[1])
        self.stride = (int(stride[0]), int(stride[1]))
        self.sigma = (sigma[0], sigma[1]) if sigma else None

    def perturb(self, ref_image: np.ndarray) -> np.ndarray:
        stride_h, stride_w = self.stride
        img_h, img_w = ref_image.shape[:2]
        center_xs = np.arange(0, img_w, stride_w)
        center_ys = np.arange(0, img_h, stride_h)

        num_masks = len(center_xs) * len(center_ys)
        masks = np.zeros((num_masks, img_h, img_w), dtype="float32" if self.sigma else "bool")
        center_xs_m = np.tile(center_xs, len(center_ys))
        center_ys_m = np.repeat(center_ys, len(center_xs))

        for i, (center_x, center_y) in enumerate(zip(center_xs_m, center_ys_m)):
            mask = masks[i]
            coords = ellipse(center_y, center_x, *self.radius, shape=mask.shape)
            mask[coords] = 1

            if self.sigma:
                mask[:] = gaussian_filter(mask, sigma=self.sigma)
                mask[:] = mask / mask.max()

            mask[:] = 1 - mask if self.sigma else ~mask

        return masks

    @classmethod
    def get_default_config(cls) -> dict[str, Any]:
        # Minor override to curry tuple defaults into lists, which are the
        # JSON-parsed types. This is to allow successful equality between
        # default, get_config() and JSON-parsed outputs.
        cfg = super().get_default_config()
        cfg["radius"] = list(cfg["radius"])
        cfg["stride"] = list(cfg["stride"])
        return cfg

    def get_config(self) -> dict[str, Any]:
        return {
            "radius": list(self.radius),
            "stride": list(self.stride),
            "sigma": list(self.sigma) if self.sigma else None,
        }
