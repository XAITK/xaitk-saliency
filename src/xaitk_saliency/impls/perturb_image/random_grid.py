from typing import Any, Dict, Optional, Tuple

import numpy as np
from skimage.transform import resize
from smqtk_descriptors.utils import parallel_map

from xaitk_saliency import PerturbImage


class RandomGrid(PerturbImage):
    """
    Generate masks using a random grid of set cell size. If the chosen cell
    size does not divide an image evenly, then the grid is over-sized and
    the resulting mask is centered and cropped. Each mask is also shifted
    randomly by a maximum of half the cell size in both x and y.

    This method is based on RISE (http://bmvc2018.org/contents/papers/1064.pdf)
    but aims to address the changing cell size, given images of different
    sizes, aspect of that implementation. This method keeps cell size constant
    and instead adjusts the overall grid size for different sized images.

    :param n: Number of masks to generate.
    :param s: Dimensions of the grid cells in pixels. E.g. (3, 4) would use
        a grid of 3x4 pixel cells.
    :param p1: Probability of a grid cell being set to 1 (not occluded).
        This should be a float value in the [0, 1] range.
    :param seed: A seed to use for the random number generator, allowing
        for masks to be reproduced.
    :param threads: Number of threads to use when generating masks. If this
        is <=0 or None, no threading is used and processing is performed
        in-line serially.
    """

    def __init__(
        self,
        n: int,
        s: Tuple[int, int],
        p1: float,
        seed: Optional[int] = None,
        threads: Optional[int] = None,
    ):
        self.n = n
        self.s = s
        self.p1 = p1
        self.seed = seed
        self.threads = threads

    def perturb(
        self,
        ref_img: np.ndarray,
    ) -> np.ndarray:
        num_masks = self.n
        s = np.array(self.s)
        p1 = self.p1
        threads = self.threads
        grid_rng = np.random.default_rng(self.seed)
        shift_rng = np.random.default_rng(self.seed)

        img_shape = np.array(ref_img.shape[:2])

        # Pad grid by one cell in both dimensions to allow for shifting
        grid_size = (img_shape // s) + 1

        # Additional padding if cell size does not divide the image evenly
        # This is required to allow shifting at full range
        grid_size += img_shape % s != 0

        grid = grid_rng.random((num_masks, *grid_size)) < p1
        grid = grid.astype(np.float32)

        masks = np.empty((num_masks, *img_shape), dtype=grid.dtype)

        up_size = grid_size * s

        # Offsets to center mask on grid
        center_y, center_x = (up_size - img_shape) // 2

        # Maximum shifting offsets
        max_shift_y, max_shift_x = s // 2

        # Expand index for efficiency
        img_h, img_w = img_shape

        # flake8: noqa
        def work_func(_i: int) -> np.ndarray:
            if max_shift_x != 0:
                shift_x = shift_rng.integers(-max_shift_x, max_shift_x)
            else:
                shift_x = 0

            if max_shift_y != 0:
                shift_y = shift_rng.integers(-max_shift_y, max_shift_y)
            else:
                shift_y = 0

            # Total offset: offset to center + random shift
            offset_x = center_x + shift_x
            offset_y = center_y + shift_y

            mask = resize(grid[_i], up_size, order=1)
            mask = mask[offset_y : offset_y + img_h, offset_x : offset_x + img_w]

            return mask

        if threads is None or threads < 1:
            for i in range(num_masks):
                masks[i, ...] = work_func(i)
        else:
            for i, m in enumerate(parallel_map(work_func, range(num_masks), cores=threads, use_multiprocessing=False)):
                masks[i, ...] = m

        return masks

    def get_config(self) -> Dict[str, Any]:
        return {
            "n": self.n,
            "s": self.s,
            "p1": self.p1,
            "seed": self.seed,
            "threads": self.threads,
        }
