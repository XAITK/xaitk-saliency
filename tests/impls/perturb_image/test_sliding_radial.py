import numpy as np
from smqtk_core.configuration import configuration_test_helper

from xaitk_saliency import PerturbImage
from xaitk_saliency.impls.perturb_image.sliding_radial import SlidingRadial


class TestRadialOcclusionBasedPerturb:
    def test_init_default(self) -> None:
        """Test empty construction since we provide defaults."""
        impl = SlidingRadial()
        assert impl.radius == (50, 50)
        assert impl.stride == (20, 20)
        assert impl.sigma is None

    def test_init_valued(self) -> None:
        """Test that constructor values pass."""
        ex_r = (123.5, 456)
        ex_stride = (444, 445)
        ex_sig = (63, 73.2)
        impl = SlidingRadial(radius=ex_r, stride=ex_stride, sigma=ex_sig)
        assert impl.radius == ex_r
        assert impl.stride == ex_stride
        assert impl.sigma == ex_sig

    def test_plugin_find(self) -> None:
        """
        This implementation has no optional plugins so it should be found and
        exposed by the super-type's impl getter.
        """
        assert SlidingRadial in PerturbImage.get_impls()

    def test_standard_config(self) -> None:
        ex_r = (123.5, 456)
        ex_stride = (444, 445)
        ex_sig = (63, 73.2)
        impl = SlidingRadial(radius=ex_r, stride=ex_stride, sigma=ex_sig)
        for inst in configuration_test_helper(impl):
            assert inst.radius == ex_r
            assert inst.stride == ex_stride
            assert inst.sigma == ex_sig

    def test_perturb_1channel(self) -> None:
        """
        Test basic perturbation on a known 1-channel image.
        Input image mode should not impact the masks output.
        """
        # Image is slightly wide, should be occluded 6-ways.
        # Simulate grayscale image (single-channel)
        white_image = np.full((4, 6), fill_value=255, dtype=np.uint8)

        impl = SlidingRadial(radius=(2, 4), stride=(2, 2))
        actual_masks = impl.perturb(white_image)

        assert np.allclose(actual_masks, EXPECTED_MASKS_4x6)

    def test_perturb_3channel(self) -> None:
        """
        Test basic perturbation on a known 3-channel image.
        Input image mode should not impact the masks output.
        """
        # Image is slightly wide, should be occluded 6-ways.
        # Simulate RGB or BGR image (3-channel)
        white_image = np.full((4, 6, 3), fill_value=255, dtype=np.uint8)

        impl = SlidingRadial(radius=(2, 4), stride=(2, 2))
        actual_masks = impl.perturb(white_image)

        assert np.allclose(actual_masks, EXPECTED_MASKS_4x6)

    def test_perturb_3channel_with_blurring(self) -> None:
        """
        Test basic perturbation on a known 3-channel image with blurring.
        Input image mode should not impact the masks output.
        """
        # Square image for uneven masking 6-ways.
        # Simulate RGB or BGR image (3-channel)
        white_image = np.full((4, 6, 3), fill_value=255, dtype=np.uint8)

        impl = SlidingRadial(radius=(2, 4), stride=(2, 2), sigma=(2, 2))
        actual_masks = impl.perturb(white_image)

        assert np.allclose(actual_masks, EXPECTED_BLURRED_MASKS_4x6)

    def test_perturb_4channel(self) -> None:
        """
        Test basic perturbation on a known 4-channel image.
        Input image mode should not impact the masks output.
        """
        # Image is slightly wide, should be occluded 6-ways.
        # Simulate RGBA format image (4-channel)
        white_image = np.full((4, 6, 4), fill_value=255, dtype=np.uint8)

        impl = SlidingRadial(radius=(2, 4), stride=(2, 2))
        actual_masks = impl.perturb(white_image)

        assert np.allclose(actual_masks, EXPECTED_MASKS_4x6)


# Common expected masks for 4x6 tests
EXPECTED_MASKS_4x6 = np.array(
    [
        [[0, 0, 0, 0, 1, 1], [0, 0, 0, 0, 1, 1], [1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]],
        [[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]],
        [[1, 0, 0, 0, 0, 0], [1, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]],
        [[1, 1, 1, 1, 1, 1], [0, 0, 0, 0, 1, 1], [0, 0, 0, 0, 1, 1], [0, 0, 0, 0, 1, 1]],
        [[1, 1, 1, 1, 1, 1], [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0]],
        [[1, 1, 1, 1, 1, 1], [1, 0, 0, 0, 0, 0], [1, 0, 0, 0, 0, 0], [1, 0, 0, 0, 0, 0]],
    ],
    dtype=bool,
)

# Common expected masks for 4x6 tests with blurring.
EXPECTED_BLURRED_MASKS_4x6 = np.array(
    [
        [
            [0.0, 0.0584771, 0.18115371, 0.35675043, 0.5382167, 0.65533626],
            [0.15234476, 0.20191318, 0.30590063, 0.45474613, 0.6085669, 0.707844],
            [0.36783165, 0.40479892, 0.48235124, 0.5933579, 0.70807517, 0.7821145],
            [0.5201763, 0.54823494, 0.6070981, 0.69135356, 0.7784254, 0.83462214],
        ],
        [
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [0.1523447, 0.1523447, 0.1523447, 0.1523447, 0.1523447, 0.1523447],
            [0.3678316, 0.3678316, 0.3678316, 0.3678316, 0.3678316, 0.3678316],
            [0.5201763, 0.5201763, 0.5201763, 0.5201763, 0.5201763, 0.5201763],
        ],
        [
            [0.36687934, 0.28730744, 0.17449367, 0.07927084, 0.02294374, 0.0],
            [0.46333194, 0.39588237, 0.30025518, 0.2195391, 0.1717931, 0.1523447],
            [0.5997611, 0.54945827, 0.47814095, 0.41794413, 0.3823359, 0.3678316],
            [0.69621366, 0.65803313, 0.60390246, 0.5582123, 0.5311852, 0.5201763],
        ],
        [
            [0.28489286, 0.32671022, 0.41443717, 0.5400076, 0.6697754, 0.75352854],
            [0.19732106, 0.2442593, 0.34272927, 0.4836771, 0.62933624, 0.7233457],
            [0.07932705, 0.13316524, 0.24611032, 0.40777743, 0.57484853, 0.6826774],
            [0.0, 0.05847704, 0.18115371, 0.35675037, 0.5382166, 0.65533626],
        ],
        [
            [0.28489286, 0.28489286, 0.28489286, 0.28489286, 0.28489286, 0.28489286],
            [0.19732106, 0.19732106, 0.19732106, 0.19732106, 0.19732106, 0.19732106],
            [0.07932705, 0.07932705, 0.07932705, 0.07932705, 0.07932705, 0.07932705],
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        ],
        [
            [0.54725087, 0.49034846, 0.40967453, 0.34158003, 0.30130005, 0.28489286],
            [0.4918074, 0.42793667, 0.3373834, 0.26095015, 0.21573746, 0.19732106],
            [0.41710287, 0.34384316, 0.23997855, 0.1523096, 0.10045069, 0.07932699],
            [0.36687934, 0.28730738, 0.17449361, 0.0792709, 0.02294368, 0.0],
        ],
    ],
    dtype="float32",
)
