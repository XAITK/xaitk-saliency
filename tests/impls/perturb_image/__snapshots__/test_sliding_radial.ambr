# serializer version: 1
# name: TestRadialOcclusionBasedPerturb.test_perturb_1channel
  array([[[False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False]],

         [[ True,  True,  True,  True,  True,  True],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False]]])
# ---
# name: TestRadialOcclusionBasedPerturb.test_perturb_3channel
  array([[[False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False]],

         [[ True,  True,  True,  True,  True,  True],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False]]])
# ---
# name: TestRadialOcclusionBasedPerturb.test_perturb_3channel_with_blurring
  array([[[0.        , 0.0584771 , 0.18115371, 0.35675043, 0.5382167 ,
           0.65533626],
          [0.15234476, 0.20191318, 0.30590063, 0.45474613, 0.6085669 ,
           0.707844  ],
          [0.36783165, 0.40479892, 0.48235124, 0.5933579 , 0.70807517,
           0.7821145 ],
          [0.5201763 , 0.54823494, 0.6070981 , 0.69135356, 0.7784254 ,
           0.83462214]],

         [[0.        , 0.        , 0.        , 0.        , 0.        ,
           0.        ],
          [0.1523447 , 0.1523447 , 0.1523447 , 0.1523447 , 0.1523447 ,
           0.1523447 ],
          [0.3678316 , 0.3678316 , 0.3678316 , 0.3678316 , 0.3678316 ,
           0.3678316 ],
          [0.5201763 , 0.5201763 , 0.5201763 , 0.5201763 , 0.5201763 ,
           0.5201763 ]],

         [[0.36687934, 0.28730744, 0.17449367, 0.07927084, 0.02294374,
           0.        ],
          [0.46333194, 0.39588237, 0.30025518, 0.2195391 , 0.1717931 ,
           0.1523447 ],
          [0.5997611 , 0.54945827, 0.47814095, 0.41794413, 0.3823359 ,
           0.3678316 ],
          [0.69621366, 0.65803313, 0.60390246, 0.5582123 , 0.5311852 ,
           0.5201763 ]],

         [[0.28489286, 0.32671022, 0.41443717, 0.5400076 , 0.6697754 ,
           0.75352854],
          [0.19732106, 0.2442593 , 0.34272927, 0.4836771 , 0.62933624,
           0.7233457 ],
          [0.07932705, 0.13316524, 0.24611032, 0.40777743, 0.57484853,
           0.6826774 ],
          [0.        , 0.05847704, 0.18115371, 0.35675037, 0.5382166 ,
           0.65533626]],

         [[0.28489286, 0.28489286, 0.28489286, 0.28489286, 0.28489286,
           0.28489286],
          [0.19732106, 0.19732106, 0.19732106, 0.19732106, 0.19732106,
           0.19732106],
          [0.07932705, 0.07932705, 0.07932705, 0.07932705, 0.07932705,
           0.07932705],
          [0.        , 0.        , 0.        , 0.        , 0.        ,
           0.        ]],

         [[0.54725087, 0.49034846, 0.40967453, 0.34158003, 0.30130005,
           0.28489286],
          [0.4918074 , 0.42793667, 0.3373834 , 0.26095015, 0.21573746,
           0.19732106],
          [0.41710287, 0.34384316, 0.23997855, 0.1523096 , 0.10045069,
           0.07932699],
          [0.36687934, 0.28730738, 0.17449361, 0.0792709 , 0.02294368,
           0.        ]]], dtype=float32)
# ---
# name: TestRadialOcclusionBasedPerturb.test_perturb_4channel
  array([[[False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True],
          [ True,  True,  True,  True,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True],
          [False, False, False, False,  True,  True]],

         [[ True,  True,  True,  True,  True,  True],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False],
          [False, False, False, False, False, False]],

         [[ True,  True,  True,  True,  True,  True],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False],
          [ True, False, False, False, False, False]]])
# ---
